### ISSUE TITLE
Symtab should not always demangle every function name

### ISSUE THREAD

**Issue by mwkrentel** [dyninst/dyninst#577]
When Symtab demangles symbol names in `P_cplus_demangle()` in
`common/src/linuxKludges.C`, it looks like Symtab automatically calls
`__cxa_demangle()` to demangle every name, whether it is mangled or
not.  This is not correct.

Instead, Symtab should lookup what module the name is from and
demangle if the source language is C++, ie, `lang_CPlusPlus` or
`lang_GnuCPlusPlus` in module-speak.  (I don't know what
`lang_Fortran_with_pretty_debug` is, but I'm guessing it isn't
mangled.)

I think the infrastructure for this is already in place, so hopefully
this is a fairly small fix.

----------

Here's how to reproduce the problem with a .c program.

```
#include <stdio.h>

void a(void)
{
    printf("hello\n");
}

int main(int argc, char **argv)
{
    a();
    return 0;
}
```

Compile this with a C compiler (not C++) without optimization or -O0
so the compiler doesn't inline "a" and eliminate it.

Then, see what Symtab produces for `mangled_names_begin()`,
`pretty_names_begin()` and `typed_names_begin()` for function "a".
I used debug output from hpcstruct for this, but by whatever means,
what I get is:

```
mangled name:  a
pretty name:   signed char
typed name:    signed char
```

The problem, of course, is that `__cxa_demangle("a") --> "signed char"`.
This is a valid demangling (of a type encoding, not a symbol name),
but in the wrong context.  The source file is C, so the symbol is not
mangled.

ping @mxz297

**Comment by mxz297** [comment]
@mwkrentel The problem is reproduced on my end. Thanks for the example. I will fix it soon.

**Comment by mxz297** [comment]
@mwkrentel @jmellorcrummey After more thoughts, I feel this problem is more complicated and I don't think there is a perfect solution for it. 

## Source language of the module did not really help 

Let me start with Mark's suggested solution, which is to look at the source language of the module. 

1. The binary is not guaranteed to be compiled with `-g`. In this case, symtabAPI says that the source language is `lang_Unknown`. 

2. Even if the binary is compiled with `-g`. Module level information is not sufficient for dealing with `extern "C"`. I modified Mark's example as follows:

```
#include <stdio.h>

extern "C" {

void a()
{
    printf("hello\n");
}

}

int main(int argc, char **argv)
{
    a();
    return 0;
}
```

SymtabAPI tells me that

```
Symbol at 400507
	getMangledName(): a
	getPrettyName(): signed char
	getTypedName(): signed char
	Lanuage: 3
```

The language enum 3 is `lang_CPlusPlus`. Obviously, the module itself is c++, but the symbol is C symbol. By using the module source language information, we cannot really determine whether we should demangle the name or not.

## We cannot distinguish C name from C++ name

Consider the following example: 

```
void a()
{
    printf("hello\n");
}

int main(int argc, char **argv)
{
    a();
    return 0;
}
```

When this example is compiled with `g++`, function `a` is mangled as `_Z1av`. We should demangle `_Z1av` for users.

Now let's see another example:

```
extern "C" {

void _Z1av()
{
    printf("hello c function\n");
}

}

int main(int argc, char **argv)
{
    _Z1av();
    return 0;
}
```

When this example is compiled with `g++`, function `_Z1av` is not mangled. But it looks like exactly the same as if function `a` is mangled. In this case, it looks like we should not demangle `_Z1av`, but how can we distinguish this example from the above one?

Just to confirm that there is no fundamental difference between C and C++ names. I create the following example:

```
#include <stdio.h>

extern "C" {

void _Z1av()
{
    printf("hello c function\n");
}

}

void a()
{
    printf("hellp c++ function\n");
}

int main(int argc, char **argv)
{
    _Z1av();
    a();
    return 0;
}
```

And I got the following error message:
```
[xmeng@mic CodeDump]$ g++ -g -o name-test name-test2.cpp
/tmp/cc6DPENK.s: Assembler messages:
/tmp/cc6DPENK.s:37: Error: symbol `_Z1av' is already defined
```

## Conclusion

My conclusion is that we can only do a best-effort to decide whether we should or should not demangle a name. 

My proposal is that HPCToolkit queries `getTypedName()` first to get the demangled name. If the demangled name does not make sense, HPTToolkit uses `getMangledName()` instead.

For example, I noticed that `_Z1av` is returned as `a()`. The key here is the parenthesis. Since C++ mangles function name to support function overload, it will encode function parameters including the parenthesis. If `getTypedName()` does not contain any parenthesis or does not have a matched pair of parenthesis, we say we wrongly demangle a C name. 

And yes, in this way, we hope that a user does not name their C functions in a way like how C++ mangles C++ functions...

**Comment by jmellorcrummey** [comment]
I think that this could be fixed in Dyninst in getTypedName.

I tried to demangle every single letter and number. Here is what I got. Based on my results, it seems that if the output contains a space but not a parenthesis, then the demangled result is bogus and the mangled name should be returned instead.
```

a -> signed char signed char 
b -> bool bool 
c -> char char 
d -> double double 
e -> long double long double 
f -> float float 
g -> __float128 __float128 
h -> unsigned char unsigned char 
i -> int int 
j -> unsigned int unsigned int 
k ->  (null) 
l -> long long 
m -> unsigned long unsigned long 
n -> __int128 __int128 
o -> unsigned __int128 unsigned __int128 
p ->  (null) 
q ->  (null) 
r ->  (null) 
s -> short short 
t -> unsigned short unsigned short 
u ->  (null) 
v -> void void 
w -> wchar_t wchar_t 
x -> long long long long 
y -> unsigned long long unsigned long long 
z -> ... ... 
A ->  (null) 
B ->  (null) 
C ->  (null) 
D ->  (null) 
E ->  (null) 
F ->  (null) 
G ->  (null) 
H ->  (null) 
I ->  (null) 
J ->  (null) 
K ->  (null) 
L ->  (null) 
M ->  (null) 
N ->  (null) 
O ->  (null) 
P ->  (null) 
Q ->  (null) 
R ->  (null) 
S ->  (null) 
T ->  (null) 
U ->  (null) 
V ->  (null) 
W ->  (null) 
X ->  (null) 
Y ->  (null) 
Z ->  (null) 
0 ->  (null) 
1 ->  (null) 
2 ->  (null) 
3 ->  (null) 
4 ->  (null) 
5 ->  (null) 
6 ->  (null) 
7 ->  (null) 
8 ->  (null) 
9 ->  (null)
```

**Comment by jmellorcrummey** [comment]
My test program:

`#include <cxxabi.h>
#include <stdio.h>


void
demangle_range(char first, int len)
{
  char in[2];
  char out[1000];
  size_t outlen;
  int status;

  for (int i = 0; i < len; i++) {
    in[0] = first + i;
    in[1] = 0;
    out[0] = 0;
    outlen = sizeof(out);
    char *result = abi::__cxa_demangle(in, out, &outlen, &status);
    if (result) printf("'%s' -> '%s'\n", in, out, result);
  }

}

int main(int argc, char **argv)
{
  char first_printable_ascii = ' ';
  int num_printable = 127-32;
  demangle_range(32, num_printable);
  return 0;
}`

The output:

```
'a' -> 'signed char'
'b' -> 'bool'
'c' -> 'char'
'd' -> 'double'
'e' -> 'long double'
'f' -> 'float'
'g' -> '__float128'
'h' -> 'unsigned char'
'i' -> 'int'
'j' -> 'unsigned int'
'l' -> 'long'
'm' -> 'unsigned long'
'n' -> '__int128'
'o' -> 'unsigned __int128'
's' -> 'short'
't' -> 'unsigned short'
'v' -> 'void'
'w' -> 'wchar_t'
'x' -> 'long long'
'y' -> 'unsigned long long'
'z' -> '...'
```

If the input does not equal the output and does not contain a parenthesis, then it must be wrong.

**Comment by mxz297** [comment]
According to the Itanium C++ ABI (https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling), the rule of mangled name is:
```
    <mangled-name> ::= _Z <encoding>
                   ::= _Z <encoding> . <vendor-specific suffix>
    <encoding> ::= <function name> <bare-function-type>
	       ::= <data name>
	       ::= <special-name>
```

So, if the name does not start with "_Z",  `getTypedName` and `getPrettyName` should just return the name of the symbol without doing any demangling. 

If the symbol starts with "_Z",  we perform the demangling, and do our best effort to check it is a legit demangled name, such as whether it contains parenthesis. If not, we still return the original symbol name.

**Comment by mwkrentel** [comment]
I'm not sure where this leaves us.  I agree that we can't achive 100%
correctness and have to rely on heuristics.  But I still suggest that
Symtab should cover the common cases and do the best it can with
whatever info is available.  Is that what we're thinking?

In particular, for profiling and debugging, we tell users to use `-g`.
If they don't, then they're going to miss other things even more
important than the correct demangled names.

**Comment by mxz297** [comment]
> I'm not sure where this leaves us. I agree that we can't achive 100%
> correctness and have to rely on heuristics. But I still suggest that
> Symtab should cover the common cases and do the best it can with
> whatever info is available. Is that what we're thinking?

Yes, that's why I am proposing following the name mangling rule to determine whether SymtabAPI should or should not demangle the name.

> 
> In particular, for profiling and debugging, we tell users to use `-g`.
> If they don't, then they're going to miss other things even more
> important than the correct demangled names.

In my long post, I have a specific section to discuss why `-g` is not helpful.

**Comment by jmellorcrummey** [comment]
I thought that we could all agree on what Xiaozhu suggested last and repeated below:

```
According to the Itanium C++ ABI (https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling), the rule of mangled name is:

    <mangled-name> ::= _Z <encoding>
                   ::= _Z <encoding> . <vendor-specific suffix>
    <encoding> ::= <function name> <bare-function-type>
	       ::= <data name>
	       ::= <special-name>
So, if the name does not start with "_Z", getTypedName and getPrettyName should just return the name of the symbol without doing any demangling.

If the symbol starts with "_Z", we perform the demangling, and do our best effort to check it is a legit demangled name, such as whether it contains parenthesis. If not, we still return the original symbol name.
```

**Comment by hainest** [comment]
As a sanity check, c++filt does not have this issue.

> $ echo a | c++filt 
> a

I took a look at the source for c++filt, and it _always_ calls `cplus_demangle` from libiberty. Bizarrely, Dyninst does not support using libiberty on linux (it does on several other platforms, though). I would recommend a combination of the "_Z" trick with the assumption that Itanium is being used even on non-x86 platforms (here be dragons!), then add a conditional check for libiberty and use that, if present.

NB: From the libiberty source: 

> Note that any leading underscores, or other such characters prepended by
> the compilation system, are presumed to have already been stripped.

**Comment by mwkrentel** [comment]
c++filt and libiberty both have options for this.  I guess the default
must be not to decode type encodings.

```
$ c++filt a
a

$ c++filt -t a
signed char
```

These are the options available for libiberty `cplus_demangle`
from `demangle.h`.

```
/* Options passed to cplus_demangle (in 2nd parameter). */

#define DMGL_NO_OPTS     0              /* For readability... */
#define DMGL_PARAMS      (1 << 0)       /* Include function args */
#define DMGL_ANSI        (1 << 1)       /* Include const, volatile, etc */
#define DMGL_JAVA        (1 << 2)       /* Demangle as Java rather than C++. */
#define DMGL_VERBOSE     (1 << 3)       /* Include implementation details.  */
#define DMGL_TYPES       (1 << 4)       /* Also try to demangle type encodings.  */
#define DMGL_RET_POSTFIX (1 << 5)       /* Print function return types (when
                                           present) after function signature.
                                           It applies only to the toplevel
                                           function type.  */
#define DMGL_RET_DROP    (1 << 6)       /* Suppress printing function return
                                           types, even if present.  It applies
                                           only to the toplevel function type.
                                           */
#define DMGL_AUTO        (1 << 8)
#define DMGL_GNU         (1 << 9)
#define DMGL_LUCID       (1 << 10)
#define DMGL_ARM         (1 << 11)
#define DMGL_HP          (1 << 12)       /* For the HP aCC compiler;
                                            same as ARM except for
                                            template arguments, etc. */
#define DMGL_EDG         (1 << 13)
#define DMGL_GNU_V3      (1 << 14)
#define DMGL_GNAT        (1 << 15)
#define DMGL_DLANG       (1 << 16)
#define DMGL_RUST        (1 << 17)      /* Rust wraps GNU_V3 style mangling.  */
```

And the **really** bizarre thing about Dyninst and libiberty is that
on Linux at least, cmake requires libiberty and the build fails without it.
But then, you don't actually use it.  :-)   Older Dyninst did use it, at least
in some cases, but 10.0 doesn't.

**Comment by hainest** [comment]
>And the really bizarre thing about Dyninst and libiberty is that
on Linux at least, cmake requires libiberty and the build fails without it.

This doesn't appear to be the case anymore. If you don't specify USE_GNU_DEMANGLER (default ON using ccmake, default OFF using cmake- I have a fix coming in for this), then it checks the system for libiberty or builds from source, if not found. In either case, we have libiberty, so the real question would be can we just get rid of USE_GNU_DEMANGLER? I'm not ready to make that decision, but the behavior of other parts of Dyninst doing demangling point to "yes."

**Comment by mwkrentel** [comment]
If you decide to use libiberty, then one thing you could do is use a
full set of options for the typed name, something like
`DMGL_PARAMS | DMGL_ANSI | DMGL_RET_POSTFIX` and then a 
reduced set of options for the pretty name, maybe `DMGL_RET_DROP`.

You could use that to replace de-demangle.
For example,

```
mangled:  _Z14CalcElemVolumeddddd
pretty:   CalcElemVolume
typed:    CalcElemVolume(double, double, double, double, double)

mangled:  _ZN4RAJA6forallIZ13CalcQForElemsP6DomainEUliE_EEvNS_21omp_parallel_for_execERKNS_9RangeISetET_._omp_fn.36
pretty:   RAJA::forall<CalcQForElems(Domain*)::{lambda(int)#1}>
typed:    RAJA::forall<CalcQForElems(Domain*)::{lambda(int)#1}>(RAJA::omp_parallel_for_exec, RAJA::RangeISet const&, CalcQForElems(Domain*)::{lambda(int)#1})void [clone ._omp_fn.36]
```

**Comment by mxz297** [comment]
Fix for this issue is in #592
